### ISSUE TITLE
Function ctm.get_thetas tales very long time to evaluate from 100K set.

### ISSUE THREAD

**Issue by bhardwaj1230** [MilaNLProc/contextualized-topic-models#18]
Heelo,

I have used the below method to work on a text documents to evaluate the topics, code works well on 100 lines of documents. But, I am facing issues when I run on 100K lines, the traning completes on time but during topic extraction, ctm.get_thetas takes way too much time (80+ hours). Is there anything I am missing here.
Thank you in advance.

#FYI:

Yes, it will predict **english** topics for the spanish unseen documents (see the paper for more details). However, if the context of the training document is completely different from the ones used at test time (e.g., you train on sport related tweet in english and test on politics related tweets in spanish) the topic prediction will be less accurate.

A few more notes:

You will need to use multilingual embeddings in this case

```python
training_bert = bert_embeddings_from_file("documents.txt", "distiluse-base-multilingual-cased")
````
and you must use the "contextual" model in place of the "combined" one:

```python
ctm = CTM(input_size=len(handler.vocab), bert_input_size=512, inference_type="contextual", n_components=50)
```
The multilingual notebook in the example folder should contain all that is needed to run this process. 

At test time, you can simply create the new vector embeddings for a different language and predict the topics. You'll find that to predict topic you can use the very same code I've shared with you in the first message.

```python
testing_bert_italian = bert_embeddings_from_file('unpreprocessed_docs_italian.txt', "distiluse-base-multilingual-cased")
testing_dataset_italian = CTMDataset(testing_bert_italian, testing_bert_italian, []) # not the best definition, but it's an effective workaround to get the topics for italian documents.
```

and you get the topic distributions:

```python
predicted_topics = [] 
thetas = np.zeros((len(testing_dataset_italian), num_topics))
for a in range(0, 100):
    thetas = thetas + np.array(ctm.get_thetas(testing_dataset_italian))
    
for idd in range(0, len(testing_dataset_italian)):
    thetas[idd] = thetas[idd]/np.sum(thetas[idd])
    predicted_topic = np.argmax(thetas[idd]) 
    predicted_topics.append(predicted_topic)

thetas # document-topic distribution 
predicted_topics # list of the topic predicted for each testing document
```

Hope this helps :)

_Originally posted by @vinid in https://github.com/MilaNLProc/contextualized-topic-models/issues/4#issuecomment-644705518_

**Comment by vinid** [comment]
Hello @bhardwaj1230!

Which version of the package are you using?

I'm assuming you are using the latest version. In version 1.5.0 of the package we have included the sampling in the actual method (see [here](https://github.com/MilaNLProc/contextualized-topic-models/blob/master/contextualized_topic_models/models/ctm.py#L238)). By default, this method samples 20 times from the distribution and then averages. Using the old version of the code, you are doing this operation 100 times more. This should greatly reduce the computation time.

Just to be clearer,  the code you see here:

```python
predicted_topics = [] 
thetas = np.zeros((len(testing_dataset_italian), num_topics))
for a in range(0, 100):
    thetas = thetas + np.array(ctm.get_thetas(testing_dataset_italian))
    
for idd in range(0, len(testing_dataset_italian)):
    thetas[idd] = thetas[idd]/np.sum(thetas[idd])
    predicted_topic = np.argmax(thetas[idd]) 
    predicted_topics.append(predicted_topic)

thetas # document-topic distribution 
predicted_topics # list of the topic predicted for each testing document
```

has been wrapped inside the method ctm.get_thetas(). In fact, ctm.get_thetas accepts a new parameter that is n_sample (see the [doc](https://contextualized-topic-models.readthedocs.io/en/latest/ctm.html)) that allows you to decide how many times to sample from the distribution before averaging.

However, I'll try to see if I can speed up this process in some way. 

Hope this helps, I'll be happy if you could provide more feedback

**Comment by bhardwaj1230** [comment]
Hello,

Thanks you for the reply.
I think I understood the part, so I am just running  ctm.get_predicted_topics(training_dataset, n_samples=20) and I feel it's bit faster.
Few more questions, 

- I am trying to predict different topics for each paragraphs and I have 300K paragraphs and I was thinking can I use any multilingual bert models ? like XLMRoBERTa or XLM ? The distiluse-base-multilingual-cased model isn't giving me the topics I am expecting.

- For now my loss values are close to 336, is there a way to reduce the loss. I mean any parameters I can tune. I am doing the preprocessing step you have mentioned.

Training Code and Log snipets :

Code : 

from contextualized_topic_models.models.ctm import CTM
from contextualized_topic_models.utils.data_preparation import TextHandler
from contextualized_topic_models.utils.data_preparation import bert_embeddings_from_file
from contextualized_topic_models.datasets.dataset import CTMDataset

handler = TextHandler('pre_processed.txt')
handler.prepare() # create vocabulary and training data

num_topics = 10
num_epochs=10
inference_type='combined' #contextual

training_bert = bert_embeddings_from_file('pre_processed.txt', "distiluse-base-multilingual-cased")

training_dataset = CTMDataset(handler.bow, training_bert, handler.idx2token)

ctm = CTM(input_size=len(handler.vocab), bert_input_size=512, num_epochs=num_epochs, inference_type=inference_type, n_components=num_topics)
ctm.fit(training_dataset) # run the model

Logs : 

Settings: 
                   N Components: 10
                   Topic Prior Mean: 0.0
                   Topic Prior Variance: 0.9
                   Model Type: prodLDA
                   Hidden Sizes: (100, 100)
                   Activation: softplus
                   Dropout: 0.2
                   Learn Priors: True
                   Learning Rate: 0.002
                   Momentum: 0.99
                   Reduce On Plateau: False
                   Save Dir: None
Epoch: [1/10]	Samples: [127307/1273070]	Train Loss: 346.56960815946513	Time: 0:00:33.557321
Epoch: [2/10]	Samples: [254614/1273070]	Train Loss: 337.93411090844864	Time: 0:00:33.506455
Epoch: [3/10]	Samples: [381921/1273070]	Train Loss: 337.38634942562567	Time: 0:00:33.707093
Epoch: [4/10]	Samples: [509228/1273070]	Train Loss: 337.2601060855998	Time: 0:00:33.673359
Epoch: [5/10]	Samples: [636535/1273070]	Train Loss: 337.2131043580767	Time: 0:00:33.686936
Epoch: [6/10]	Samples: [763842/1273070]	Train Loss: 337.2483788688159	Time: 0:00:33.616638
Epoch: [7/10]	Samples: [891149/1273070]	Train Loss: 336.97708037759793	Time: 0:00:33.662965
Epoch: [8/10]	Samples: [1018456/1273070]	Train Loss: 337.0389980028592	Time: 0:00:33.552014
Epoch: [9/10]	Samples: [1145763/1273070]	Train Loss: 337.1478314508678	Time: 0:00:33.676008
Epoch: [10/10]	Samples: [1273070/1273070]	Train Loss: 336.9242249381355	Time: 0:00:33.495577

**Comment by vinid** [comment]
Hello!

I guess the loss decrease is due to different factors, but I'm not sure about the setup of your training process, could you add more details?

I'm assuming you are trying to do cross lingual topic modeling, but if I'm wrong tell me :) :

- If you want to do cross-lingual topic modeling you cannot use the **combined** model, but you need to use the **contextual** one. 

- Remember that the training data should, in general, come from only one language. This is because we are going to keep the output BoW of one language as a base even for the languages. You can refer to the [paper](https://arxiv.org/abs/2004.07737) for more details about this. In the paper we train on English and test on Italian, Portuguese, French and German. Also, in the paper we used **distilbert-base-nli-mean-tokens**.

- Which languages are you using? Are they covered by **distiluse-base-multilingual-cased**?

- `training_bert = bert_embeddings_from_file('pre_processed.txt', "distiluse-base-multilingual-cased")`, you might want to use the non preprocessed documents here as they might help sbert in giving better representations.

In the meantime, I'll give a look to the get_thetas method to check if there are problems with that!

**Comment by bhardwaj1230** [comment]
Hello,

Thanks for the quick reply,

- If you want to do cross-lingual topic modeling you cannot use the combined model, but you need to use the contextual one.

Yes the data is mixed, both English and French. If this setting dosen't work then, can I just train on English and test on both ?

- Remember that the training data should, in general, come from only one language. This is because we are going to keep the output BoW of one language as a base even for the languages. You can refer to the paper for more details about this. In the paper we train on English and test on Italian, Portuguese, French and German.

Thank you, I will read it indetail. 

- Which languages are you using? Are they covered by distiluse-base-multilingual-cased?

Both English and French. The data is in pragraphs formate store in a text files. Ench praragraph can be either English or French.

- training_bert = bert_embeddings_from_file('pre_processed.txt', "distiluse-base-multilingual-cased"), you might want to use the non preprocessed documents here as they might help sbert in giving better representations.

Ok I will train on un-processed data with contextual setting and get back to you. Although the predicted_topics = ctm.get_predicted_topics(training_dataset, n_samples=100) function is still slow for larger data. It works fast with 10K praragraphs, but not when I try to get topics for all 300K praragraphs.


Thank you again.

**Comment by vinid** [comment]
Hi!

thanks for the feedback on the get_thetas problem, it might be just a batching issue. I'll check this a bit more in detail.

For the other questions: yes, you can train on English documents and then predict on French (and English) documents (this is very similar to the setting we have used in the paper). You will get, in output, English topics for French documents. Remember to use the contextual model and not the combined one.

Another thing you might want to consider is to check the preprocessing. We have generally found good results when limiting the number of tokens to 2000, for the pre-processed documents. We have a very simple preprocessing pipeline that should allow you to quickly clean your documents in `contextualized_topic_models.utils.preprocessing` (you will also find it mentioned in the README file). This might also reduce the training time.

Happy to add more details if needed :)

**Comment by bhardwaj1230** [comment]
Hello Again :)

Is it possible to save the model once train ? SInce I am doing it on Colab, I have to train it everytime.
Thanks.

**Comment by vinid** [comment]
Hello! 

You should be able to save the trained model with `torch.save(model, PATH)` (see this [PyTorch Tutorial](https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-loading-model-for-inference)); I have just tested this and it seems to work, however as the authors of the tutorial page describe, this solution might come with some possible downsides.

**Comment by vinid** [comment]
I'm closing this issue :) feel free to open another one in case you need :)
